### PR TITLE
improve the suggestion of the lint `unit-arg`

### LINKED ISSUE #5823: unit_arg suggestions don't make sense when run against Option::map_or(unit_expr, ...)
```rust
#[deny(clippy::unit_arg)]
pub fn foo(o: Option<i32>, f: impl FnOnce(i32), g: impl FnOnce()) {
    o.map_or(g(), |i| f(i))
}
```

(For context, this is the code after taking the code in #5821 and applying `clippy::option_if_let_else`'s suggestion.)

The lint's suggestion is:

```
help: move the expression in front of the call...
  |
3 |     g();
  |
help: ...and use a unit literal instead
  |
3 |     o.map_or((), |i| f(i))
  |              ^^
```

I'm not sure what the first suggestion is trying to say.

The second suggestion is removing the call to `g` entirely, which is definitely wrong.

Note that the expectation of this code was that it would actually trigger `clippy::or_fun_call` to suggest `o.map_or_else(|| g(), |i| f(i))` (as mentioned in #5821), so the problem is not only that it ended up triggering `clippy::unit_arg` instead but also that the suggestions don't make sense.

### Meta

- `cargo clippy -V`: `clippy 0.0.212 (39d5a61 2020-07-17)`
- `rustc -Vv`:
  ```
  rustc 1.47.0-nightly (39d5a61f2 2020-07-17)
  binary: rustc
  commit-hash: 39d5a61f2e4e237123837f5162cc275c2fd7e625
  commit-date: 2020-07-17
  host: x86_64-unknown-linux-gnu
  release: 1.47.0-nightly
  LLVM version: 10.0
  ```

### LINKED ISSUE #6015: false positives and crazy suggestsiong for unit_arg lint
I have this code:

```rust
    pub(crate) fn set_word(&mut self, addr: Word, value: Word) -> ExcResult<()> {
        Ok(self.ram.set_word(addr, value))
    }
```

ExcResult is Result<(), Exception>

 clippy says 

```
475 |         Ok(self.ram.set_word(addr, value))
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[warn(clippy::unit_arg)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unit_arg
help: move the expression in front of the call...
    |
475 |         self.ram.set_word(addr, value);
    |
help: ...and use a unit literal instead
    |
475 |         Ok(())
    |            ^^
```

which is clearly not right. 

I have half a dozen similar errors when calling functions that return ()
### Meta

clippy 0.0.212 (04488afe3 2020-08-24)
cargo 1.46.0 (149022b1d 2020-07-17)
rustc 1.46.0 (04488afe3 2020-08-24)

### PR DESCRIPTION
**Pull request by montrivo**
Fixes #5823
Fixes #6015 

Changes 
```
help: move the expression in front of the call...
  |
3 |     g();
  |
help: ...and use a unit literal instead
  |
3 |     o.map_or((), |i| f(i))
  | 
```
into 
```
help: move the expression in front of the call and replace it with the unit literal `()`
  |
3 |     g();
  |     o.map_or((), |i| f(i))
  |
```
changelog: improve the suggestion of the lint `unit-arg`

### CODE REVIEW

**Comment by flip1995** [comment]
r? @flip1995 

Since I produced the latest (problematic) suggestion.

File: tests/ui/unit_arg.stderr [hunk 1]
```diff
@@ -147,35 +121,56 @@ help: remove the semicolon from the last statement in the block
...
-LL |         foo(0);
-LL |         foo(1);
-LL |     };
-LL |     {
-LL |         foo(2);
+LL |             foo(0);
+LL |             foo(1);
+LL |         }; {
+LL |             foo(2);
+LL |             foo(3);
```
Comment (flip1995) [inline tests/ui/unit_arg.stderr:131]: This looks weird

File: tests/ui/unit_arg.stderr [hunk 2]
```diff
@@ -147,35 +121,56 @@ help: remove the semicolon from the last statement in the block
...
+
+error: use of `or` followed by a function call
+  --> $DIR/unit_arg.rs:51:10
    |
-LL |         (),
-LL |         (),
+LL |     None.or(Some(foo(2)));
+   |          ^^^^^^^^^^^^^^^^ help: try this: `or_else(|| Some(foo(2)))`
    |
+   = note: `-D clippy::or-fun-call` implied by `-D warnings`
```
Comment (flip1995) [inline tests/ui/unit_arg.stderr:140]: Just allow this lint

File: tests/ui/unit_arg.rs [hunk 3]
```diff
@@ -47,6 +47,11 @@ fn bad() {
             foo(3);
         },
     );
+    // here Some(foo(2)) isn't the top level statement expression, wrap the suggestion in a block
+    None.or(Some(foo(2)));
```
Comment (flip1995) [inline tests/ui/unit_arg.rs:51]: Great catch!

**Review by flip1995 — COMMENTED** [review]
Thanks for further improving the suggestion! Only one case is indented weirdly now.

File: tests/ui/unit_arg.stderr (hunk 2 above)
Comment (montrivo) [reply tests/ui/unit_arg.stderr]: Still a bit weird, I must still handle this.

File: tests/ui/unit_arg.stderr (hunk 1 above)
Comment (montrivo) [reply tests/ui/unit_arg.stderr]: There is still an indentation issue I must handle.

Comment (montrivo) [reply tests/ui/unit_arg.stderr]: Fixed

File: clippy_lints/src/types.rs [hunk 4]
```diff
@@ -856,18 +857,38 @@ fn lint_unit_args(cx: &LateContext<'_>, expr: &Expr<'_>, args_to_recover: &[&Exp
...
+                    .map(|(i, v)| {
+                        let with_indent_prefix = if i > 0 { " ".repeat(stmts_indent) + &v } else { v };
+                        trim_multiline(with_indent_prefix.into(), true, Some(stmts_indent)).into_owned()
+                    })
+                    .collect::<Vec<String>>()
+                    .join(";\n");
+
+                if wrap_in_block {
+                    stmts_and_call_str = " ".repeat(stmts_indent) + &stmts_and_call_str;
+                    stmts_and_call_str = format!("{{\n{}\n{}}}", &stmts_and_call_str, " ".repeat(indent));
```
Comment (flip1995) [inline clippy_lints/src/types.rs:887]: Would it make sense to wrap this `wrap_in_block` handling code into its own function to get rid of the `allow` above?

**Review by flip1995 — APPROVED** [review]
LGTM overall, except one code-style thing.

File: clippy_lints/src/types.rs (hunk 4 above)
Comment (montrivo) [reply clippy_lints/src/types.rs]: I moved the that logic out. That clarifies the code a bit :+1:.
I had to modify and rename the `trim_multiline` utils method to let it indent code further as it only supported dedentation.

Comment (montrivo) [reply clippy_lints/src/types.rs]: in case you missed the notif @flip1995 (sorry if have no time atm)

**Comment by flip1995** [comment]
@bors r+

Thanks! And sorry for the late reply

**Pull request merged**